Revert "Drop MRI 1.8.7-2.2 support (#158)"

pboling opened the pull request:
This reverts commit 92d5134994003d6f17e874117483bac7e8fb7829.

Also fixes a few bugs.

- [x] indentation was inconsistent across various BundlerDSL types
- [x] Ruby <= 2.7 will never work with gems sourced from git on local filesystem, because old bundler (2.4.22) doesn't support it.
- [x] Adds Ruby 2.7, 3.4 to CI, and they are passing
- [x] Fixes CI for existing builds for Ruby 3.0, 3.1, 3.2, 3.3, ruby-head 
- [x] Adds Ruby 2.6 to CI, and it is failing, as will any lower Ruby version.  I propose we fix them in a separate PR so the CI is in a better state on the main branch.
- [x] Independently verified that these changes result in compatibility with every MRI Ruby 1.9.3+, and JRuby 9.1+ in a gem I [pinned to use a branch on my fork](https://github.com/pboling/rspec-pending_for) that integrates these changes with the `eval_gemfile` feature.

Also fixes other CI consistency issues, and adds lint / style enforcement for Ruby 1.8+ via [rubocop-lts](https://rubocop-lts.gitlab.io/).

<sub>

Please help me support open source by [sponsoring me](https://github.com/sponsors/pboling) - @pboling

</sub>

pboling committed:
📝 Documentation

pboling committed:
Revert "Drop MRI 1.8.7-2.2 support (#158)"

This reverts commit 92d5134994003d6f17e874117483bac7e8fb7829.

n-rodriguez commented on spec/support/dependency_helpers.rb:
```diff
@@ -52,11 +52,12 @@ def build_git_gem(gem_name, version = "1.0.0")
-      `git config user.name "Appraisal"`
-      `git add .`
-      `git commit --all --no-verify --message "initial commit"`
+      %x[git init . --initial-branch=master]
```
Can you please use %x()? It reads better than %x[] or %x{}. Thank you!

n-rodriguez commented on spec/support/dependency_helpers.rb:
```diff
@@ -23,23 +24,22 @@ def build_gem(gem_name, version = "1.0.0")
-        puts "building gem: #{gem_name} #{version}" if ENV["VERBOSE"]
-        `gem build #{gemspec} #{redirect}`
+      puts "building gem: #{gem_name} #{version}" if ENV["VERBOSE"]
+      %x[gem build #{gemspec} #{redirect}]
```
Can you please use %x()? It reads better than %x[] or %x{}. Thank you!

n-rodriguez commented on bin/bundle:
No need to lint this file : it is generated and about to disappear https://github.com/rubygems/rubygems/pull/8345

pboling replied:
👍 the complication is the linting is automated when running the default rake task now, and violations will clog up the new lint lockfile.  This will encourage us to get rid of files we don't use!

pboling commented on lib/appraisal/command.rb:
```diff
@@ -26,30 +26,28 @@ def run
-        end
-      end
+      ERROR
+      exit(1)
```
Wow Hound is a mess. 🤯

pboling committed:
💚 Fix bundler versions per Ruby version

- https://dev.to/galtzo/matrix-ruby-gem-bundler-etc-4kk7

n-rodriguez commented on .github/workflows/main.yml:
```diff
@@ -38,10 +38,27 @@ jobs:
+        # See: https://dev.to/galtzo/matrix-ruby-gem-bundler-etc-4kk7
         run: |
           case ${RUBY_VERSION} in
+            1.8|1.9|2.0|2.1|2.2)
```
Maybe we should add EOL Rubies in `matrix:ruby` above?

pboling replied:
Yep, that's my next task.

pboling replied:
Adding them one by one now...

n-rodriguez commented on spec/acceptance/cli/update_spec.rb:
```diff
@@ -20,8 +20,8 @@
     in_test_directory do
-      `gem uninstall dummy -v 1.0.1`
-      `gem uninstall dummy2 -a`
+      %x{gem uninstall dummy -v 1.0.1}
```
Can you please use %x()? It reads better than %x[] or %x{}. Thank you!

pboling replied:
Fixed in 2d022fcbd3f0404f07e570ca78d7e256eaddebcb

n-rodriguez commented on spec/acceptance/cli/update_spec.rb:
```diff
@@ -20,8 +20,8 @@
-      `gem uninstall dummy -v 1.0.1`
-      `gem uninstall dummy2 -a`
+      %x{gem uninstall dummy -v 1.0.1}
+      %x{gem uninstall dummy2 -a}
```
ditto

pboling replied:
Fixed in 2d022fcbd3f0404f07e570ca78d7e256eaddebcb

n-rodriguez commented:
@pboling thank you! 👍 
@nickcharlton LGTM

pboling committed:
💚 Fix ruby declaration in Gemfile when on ruby-head

- Support debug for debugging with Ruby >= 2.7
- Support byebug for debugging with Ruby < 2.7

pboling committed:
🚨 Address code review from @n-rodriguez

- https://github.com/thoughtbot/appraisal/pull/250#discussion_r1964410061

pboling committed:
👷 Add Rubies 2.7, 2.6, 2.5, 2.4, 2.3, 2.2, 2.1, 2.0, 1.9, 1.8 to build matrix

- all current jruby and truffleruby are compatible with Ruby 3.1 at minimum, thus are supported by latest bundler

pboling committed:
♻️ Use .rspec to dry spec file headers

- Use strip_heredoc uniformly in acceptance tests

n-rodriguez commented on .rspec:
```diff
@@ -0,0 +1,3 @@
+--format progress
+--color
+--require spec_helper
```
You can also add `--warnings`

pboling replied:
Done in https://github.com/thoughtbot/appraisal/pull/250/commits/a19e70746cce6bdab027e914c62a497033af181c

pboling committed:
🚨 Address code review from @n-rodriguez

- https://github.com/thoughtbot/appraisal/pull/250#discussion_r1964532654

pboling committed:
🐛 Fixes for Ruby 2.7 compat

- Differences in handling of whitespace between versions of Ruby are often seen as minor, and go unnoticed unless tests explicitly test white space as appraisal's test suite does.

pboling commented:
@n-rodriguez @nickcharlton Getting the test suite working with Ruby 2.7 was a lot of work.  The code had drifted away from compatibility with Ruby 2.7, since it wasn't in the test suite.  I've only added Ruby 2.7 and 3.4 to the matrix here, and my next task will be to add each Ruby down the line in a discrete PR, while maintaining compatibility with the newer Rubies.

The latest spec failures are all related to whitespace on empty lines within block elements of the generated appraisals, and I'm working on a solution for that so the various Rubies don't have to match there exactly.

pboling committed:
👷 add Ruby 3.4

pboling committed:
🐛 Fix inconsistent indentation

pboling committed:
📝 Documentation

pboling committed:
💚 Fix inconsistent indentation in spec fixtures

pboling committed:
🚨 Lint lock update

pboling committed:
👷 Add Ruby 2.6

pboling commented:
@n-rodriguez @nickcharlton 
OK, All Rubies are passing except for the one I just added, 2.6.  Looking at the failures, and it is at least in part due to the pattern of installing a more modern bundler without upgrading the underlying RubyGems.  Normally they should be upgraded in lock-step.  Getting past that issue will require some additional work to the CI setup, which I am very familiar with, and I have a ready made set of workflows that will accomplish it.  It is a lot of new lines though, so perhaps this should be merged, so we at least have a passing build for all MRI Rubies 2.7 - HEAD?

pboling commented on lib/appraisal/utils.rb:
```diff
@@ -7,6 +7,13 @@ def self.support_parallel_installation?
 
+    def self.support_git_local_installation?
+      Gem::Version.create(Bundler::VERSION) > Gem::Version.create("2.4.22")
+    end
```
There may be other places we'd like to integrate this check.  So far it is only used in specs.

pboling commented on spec/support/acceptance_test_helpers.rb:
```diff
@@ -43,6 +50,15 @@ module AcceptanceTestHelpers
+  #   "ruby 3.5.0dev (2025-02-20T18:14:37Z ... etc)
+  def ruby_dev_append
+    RUBY_DESCRIPTION.include?("dev") ? "-dev" : ""
+  end
```
The `head` build wouldn't pass without this hack.  The specs that test seetting the Ruby version directly in the Gemfile need that line to say `3.5.0-dev`, or Bundler will think that `3.5.0` and `3.5.0-dev` are different versions of Ruby, and exit prematurely.

pboling commented on spec/support/dependency_helpers.rb:
```diff
@@ -1,69 +1,106 @@
+      # Caller may turn it into a git repo here
+      yield build_dir, redirect if block_given?
+
+      unless skip_build
```
The git sourced gems do not need to be built or installed, so skipping those steps makes the build run a little faster.

pboling commented on spec/support/dependency_helpers.rb:
```diff
@@ -1,69 +1,106 @@
+      %x(git init . --initial-branch=main #{redirect})
+      %x(git config user.email "appraisal@thoughtbot.com" #{redirect})
+      %x(git config user.name "Appraisal" #{redirect})
+      %x(git config commit.gpgsign false #{redirect})
```
Local developers who use commit signing will need this in order for the test suite to pass.

pboling committed:
✏️ Lookahead != Lookbehind

pboling committed:
✏️ Fix typo in documentation

pboling committed:
👷 Remove Ruby 2.6 from CI (to be added back in another PR)

pboling committed:
🐛 Ruby 1.8 compat: Replace __dir__ with File.dirname(__FILE__)

pboling committed:
🚨 Lint lock update

pboling committed:
🐛 Ruby < 3.0 compat: Bifurcate the Modern Double Splat

pboling committed:
🐛 Ruby < 3.0 compat: Bifurcate the Modern Double Splat (redux)

pboling committed:
🐛 Ruby < 3.0 compat: Bifurcate the Modern Double Splat (redux deux)

pboling committed:
🐛 Ruby < 3.0 compat: Bifurcate the Modern Double Splat (tiga)

pboling committed:
🐛 Ruby < 3.0 compat: Bifurcate the Modern Double Splat (empat)

pboling committed:
🚨 Style/SymbolArray: EnforcedStyle: brackets

- Ruby < 2 compat

pboling commented:
@n-rodriguez @nickcharlton I have independently verified that these changes result in compatibility with every MRI Ruby 1.9.3+, and JRuby 9.1+ in a gem I [pinned to use a branch on my fork](https://github.com/pboling/rspec-pending_for) that integrates these changes with the `eval_gemfile` feature.

Ruby 1.8.7 can't be run any more on GHA, so we can't test it... I don't think that's a major issue.  The RSpec project is in the same situation.

I was doing that so I could release a new version of that gem (`rspec-pending_for`) specifically to use in this gem's spec suite, to skip a spec on all engines except truffleruby, since it is passing here only on trufffleruby :/

Once I do that the whole CI will be green.

pboling committed:
💚 Pend spec failure

- Possibly due to bundler issue on MRI and JRuby engines
- https://github.com/rubygems/rubygems/issues/8518

pboling committed:
💚 skip tests that are known to fail on old bundler versions

pboling committed:
💚 turn off truffleruby-head for now

pboling commented on lib/appraisal/bundler_dsl.rb:
```diff
@@ -136,8 +147,8 @@ def dependencies_entry_for_dup
 
-    %i[gits paths platforms groups source_blocks install_if].each do |method_name|
-      class_eval <<-METHODS, __FILE__, __LINE__
+    [:gits, :paths, :platforms, :groups, :source_blocks, :install_if].each do |method_name|
```
Hound does not understand Ruby < 2 compatibility at all 👯

pboling commented:
@n-rodriguez @nickcharlton This is green as a forest.  :)

pboling commented:
@nickcharlton ping!

nickcharlton commented:
Ooh, fascinating. I'm only just starting to look at this.

My initial impression (when I saw the notification weeks ago…) was that going back this far was a bit silly, but I'm impressed by how you've been able to actually make it work.

Did you have any thoughts on running the really old versions of Ruby in the tests? Do Docker images still exist that could be used?

nickcharlton commented:
I'd like to split this PR into a few chunks:

* Restoring MRI 1.8.7-2.2 support
  * this should include anything to get it to run, but not what should be in the next
* Add Rubies 2.7, 2.6, 2.5, 2.4, 2.3, 2.2, 2.1, 2.0, 1.9, 1.8 to build matrix
  * if we can, it'd be good to do this in a few stages potentially, to keep the diff as small as possible
* The main change in "Use .rspec to dry spec file headers" is really in the `strip_heredoc`, and that'd be best as it's own commit,
  * I'm less interested in the original change here, as I'm not sure it's that valuable, but I'm open to thoughts!
* Probably multiple (?) PRs for different Ruby compatibilities, but I'm less sure how they fit together as I write this

I'd like to get to one commit (and also one PR) per logical change so it's easier to track back to what commit did what. Would you mind doing that? I know it's a lot of work, so I'll make sure we can merge these in close to when you open them. You might need to stack them on many branches, so potentially it's worth doing it over a couple of days so we don't make things incredibly confusing.

pboling commented:
So, there are other projects that run tests against ancient Ruby, such as RSpec, which still tests against Ruby 1.8.7.  They use a docker image for that.

GitHub actions can build back to Ruby 1.9.3, which means getting that far back is a mostly simple path.  I didn't add it all here, because the PR was already so huge.

Regarding migrating smaller sets of Ruby compatibility, I have some thoughts.

1. This is reverting a commit that removed the compat, so I'm not sure how I would break that apart.
2. Much of the compatibility with old Ruby is automatic via the autocorrect and "style" settings in RuboCop that enforce compatibility with older Ruby.  So I could step those changes one by one, but its busy work, just PRing the autocorrect for each version of rubocop-lts.
3. I'm happy to do it, but I'll need more clarity on the path forward.

I'm going to restate, with some adjustments:

1. PR to level up CI to test against all all current supported Ruby back to 2.7 (that is _mostly_ what this PR is, from a LOC perspective)
2. PR to restore Ruby 2.3 - 2.7 Support, extending CI testing back to 2.3
4. PR to restore Ruby 1.8.7 - 2.2 Support, extending CI testing back to 1.9.3
5. PR to test Ruby 1.8.7 in CI (this is a hard maybe).

I want to get the new CI in place structurally, so it can easily incorporate the older Rubies as I add them, so that's why the CI upgrade is the first PR in the list above.

Does that sound good @nickcharlton ?  I may subdivide further if additional clean breaks present themselves, with a focus on small PRs.

I am making heavy use of my [fork + branch](https://github.com/pboling/appraisal/tree/galtzo) that has all of this plus the `eval_gemfile` support on a lot of things that I have urgent need of (most recently my [rewrite of the old](https://github.com/oauth-xx/masq2) `masq` gem as `masq2`), so it may take me a while to get to this. Perhaps not before RubyKaigi (my first ever Ruby conference!).

pboling commented:
All of this work is in [appraisal2](https://github.com/appraisal-rb/appraisal2).
